avr-hal-generic/src/serial.rs

```diff
@@ -40,7 +110,7 @@ macro_rules! impl_usart {
                 p: $USART,
                 rx: $rxmod::$RX<$crate::port::mode::Input<RX_MODE>>,
                 tx: $txmod::$TX<$crate::port::mode::Output>,
-                baud: u32,
+                baud: Baudrate<CLOCK>,
```

Comment:
I'm not 100% convinced that this is what we want to do.  I think there is value in allowing `u32` as well, which will then just go with the default implementation.  I.e. I'd allow any `Into<Baudrate<CLOCK>>` here and implement

```rust
impl<CLOCK: crate::clock::CLOCK> From<u32> for Baudrate<CLOCK> {
    ...
}
```

Reply:
An alternative would be to implement an extension trait on `u32` for converting a value into a baudrate.  E.g.

```rust
trait BaudExt {
    fn into_baud<CLOCK: crate::clock::Clock>(self) -> Baudrate<CLOCK>;
}

impl BaudExt for u32 {
    fn into_baud<CLOCK: crate::clock::Clock>(self) -> Baudrate<CLOCK> {
        Baudrate::new(self)
    }
}
```

and in application code

```rust
use arduino_leonardo::prelude::*;

Serial::new(..., 57600.into_baud());
```

Reply:
With `Into` I'm getting `the trait `Into` cannot be made into an object`

Reply:
You have to either make it a generic parameter:

```rust
pub fn new<B: Into<Baudrate<CLOCK>>>(..., baudrate: B)
```

or use the `impl Trait` syntax:

```rust
pub fn new(..., baudrate: impl Into<Baudrate<CLOCK>>)
```

---

Though, to be honest I prefer my second alternative suggestion even more.  It would allow us to provide board-specific `into_baud()` implementations so boards that need non-standard calculations can be supported easily.

Reply:
Will move to second option

Reply:
Not sure where put `BaudrateExt` trait. It should be in board lib?

Reply:
`BaudrateExt` goes into the prelude as an anonymous import:

```rust
pub use ...::BaudrateExt as _;
```

Reply:
Can you please review if I done this in correct way? Thanks.

Reply:
The import looks good but you put it in the wrong place.  Sorry, I should have explained better:  It's meant to go into the `prelude` module which is a module imported in all the examples as 

```rust
use arduino_leonardo::prelude::*;
```

This pulls all the traits into scope which means all their methods are available to be called (because a trait must be in scope for its methods to be callable).  The anonymous import

```rust
use ...::BaudrateExt as _;
```

means that the 'name' `BaudrateExt` is not actually imported, which would 'pollute' the namespace where `prelude` is imported.  You can think of this as importing `BaudrateExt` under an alternate name like `BaudrateExt32748937248972394792389479` which is unlikely to clash with any names defined in the file importing `prelude`.

Anyway, in practical terms, right now `arduino-leonardo` just reexports the prelude from the HAL:

https://github.com/Rahix/avr-hal/blob/7337cd76cd96f2d27701b137396d94a06d3a501d/boards/arduino-leonardo/src/lib.rs#L54

Instead of this, you want to create a new `prelude` module here:

```rust
pub mod prelude {
    // Expose everything from the HAL prelude
    pub use atmega32u4_hal::prelude::*;

    pub use ...::BaudrateExt as _;
}
```